cdap-tms/src/main/java/co/cask/cdap/messaging/cache/MessageCache.java

```diff
@@ -0,0 +1,513 @@
+        largestCacheEntry = largestCacheEntry == null ? cache.ceiling(cacheEntry) : largestCacheEntry;
+        if (largestCacheEntry != null && comparator.compare(largestCacheEntry.getEntry(), cacheEntry.getEntry()) >= 0) {
+          // Entries must be in strictly increasing order
+          // Clear the cache to reset state. This is just for precaution, as this shouldn't happen,
+          // unless there is bug in the TMS system (from the caller side).
+          currentWeight.addAndGet(-1 * cacheEntry.getWeight());
+          clear();
+          throw new IllegalArgumentException("Cache entry must be in strictly increasing order. " +
+                                               "Entry " + entry + " is smaller than or equal to " +
+                                               largestCacheEntry.getEntry());
+        }
+
+        // It's ok to "leak" this to reader even if the new weight is larger than the hard limit
+        // The entry will get removed eventually and the read/write operations as a whole still give valid
+        // results
+        cache.add(cacheEntry);
+        entriesAdded++;
+        largestCacheEntry = cacheEntry;
+      }
+
+      metricsContext.increment(METRICS_ADD_REQUESTS, 1L);
+      metricsContext.increment(METRICS_ENTRIES_ADDED, entriesAdded);
+      metricsContext.gauge(METRICS_WEIGHT, newWeight);
+
+      if (newWeight > limits.get().getHardLimit()) {
+        reduceWeight();
+        metricsContext.increment(METRICS_ADD_REDUCE_WEIGHT, 1L);
+      } else if (newWeight > limits.get().getReduceTrigger()) {
+        needReduceWeight.compareAndSet(false, true);
+      }
+    } finally {
+      adding.set(false);
+    }
+  }
+
+  /**
+   * Creates a {@link Scanner} for fetching cached entries in ascending order.
+   *
+   * @param startEntry the entry to start fetching from
+   * @param includeStart {@code true} to include the startEntry in the resulting {@link Scanner}
+   *                                 if it exists in the cache
+   * @param limit maximum number of entries to fetch
+   * @return a {@link Scanner} for accessing to the fetched entries
+   */
+  public Scanner<T> scan(T startEntry, boolean includeStart, int limit, MessageFilter<T> filter) {
+    List<T> entries = new LinkedList<>();
+
+    // Acquire the read lock and copy the entries. This is to guard against weight reduction while the caller
+    // is iterating using the returned Scanner.
+    cacheLock.readLock().lock();
+    T firstInCache;
+    try {
+      firstInCache = cache.isEmpty() ? null : cache.first().getEntry();
+      for (CacheEntry<T> cacheEntry : cache.tailSet(new CacheEntry<>(startEntry, 0), includeStart)) {
+        if (entries.size() >= limit) {
+          break;
+        }
+
+        MessageFilter.Result result = filter.apply(cacheEntry.getEntry());
+        if (result == MessageFilter.Result.ACCEPT) {
+          entries.add(cacheEntry.getEntry());
+        } else if (result == MessageFilter.Result.HOLD) {
```

Comment:
we will. If the result is `SKIP`, then it is skipped (by not adding to `entries`).